include/aws/s3/s3_buffer_pool.h

```diff
@@ -0,0 +1,136 @@
+#ifndef AWS_S3_BUFFER_POOL_H
+#define AWS_S3_BUFFER_POOL_H
+
+#include <aws/common/ref_count.h>
+#include <aws/io/future.h>
+#include <aws/s3/s3.h>
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+/**
+ * Generic memory pool interface.
+ * Allows consumers of s3 crt to override how buffer allocation for part buffers is done.
+ * Refer to docs/memory_aware_request_execution.md for details on how default implementation works.
+ * WARNING: this is currently experimental feature and should be used with caution.
+ * At highlevel the flow is as follows:
+ * - crt scheduler queues up requests to be prepared
+ * - requests being prepared will try to reserve mem (i.e. obtain a ticket) and wait until they get it before proceeding
+ * - once mem is reserved requests will proceed with the pipeline
+ * - request will acquire buffer from the ticket when needed
+ * - ticket is released when request is done with the buffer
+ * Note: in some cases pipeline can stall if new buffer cannot be allocated (ex. async writes flow).
+ * In this case reserve request will indicate that not granting the ticket can block and buffer pool should try to
+ * allocate ticket right away (or wait and call waker when mem is allocated for the case of async writes).
+ */
+
+AWS_PUSH_SANE_WARNING_LEVEL
+AWS_EXTERN_C_BEGIN
+struct aws_s3_buffer_ticket;
+
+/**
+ * aws_future<aws_s3_buffer_ticket*>
+ * Buffer ticket future used for reservations.
+ */
+AWS_FUTURE_T_POINTER_WITH_RELEASE_DECLARATION(aws_future_s3_buffer_ticket, struct aws_s3_buffer_ticket, AWS_S3_API)
+
+/**
+ * Meta information about ticket reservation request.
+ */
+struct aws_s3_buffer_pool_reserve_meta {
+    struct aws_s3_client
+        *client; /* client reserving the ticket. accounts for buffer pool being shared between clients. */
+    struct aws_s3_meta_request *meta_request; /* meta request ticket is being reserved for. */
```

Comment:
The `can_block` boolean name is a bit ambiguous. Consider renaming to `is_critical` or `bypass_limits` to better convey the purpose - whether this reservation should bypass normal memory limits because it's critical for operation.